app crashes on clicking send if location is disabled

#### Software and hardware versions 
Reproduced the bug in android-8.1  (Samsung j7 2016)
#### Problem description
 trying to send the form when the location is disabled crashes the app.
#### Steps to reproduce the problem
disable the location 
try to send any form
#### Expected behaviour
should ask the user to enable location


#### Error Log when the app crashes:
```2019-03-20 18:58:11.524 16900-16900/org.odk.share E/WifiHotspotHelper: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Method.invoke(Native Method)
        at org.odk.share.controller.WifiHotspotHelper.setWifiConfig(WifiHotspotHelper.java:110)
        at org.odk.share.controller.WifiHotspotHelper.disableHotspot(WifiHotspotHelper.java:133)
        at org.odk.share.activities.SendActivity.onCreate(SendActivity.java:119)
        at android.app.Activity.performCreate(Activity.java:7258)
        at android.app.Activity.performCreate(Activity.java:7249)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1222)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2927)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3059)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1724)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:7000)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:441)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1408)
     Caused by: java.lang.SecurityException: App not allowed to read or update stored WiFi AP config (uid = 10357)
        at android.os.Parcel.readException(Parcel.java:2029)
        at android.os.Parcel.readException(Parcel.java:1975)
        at android.net.wifi.IWifiManager$Stub$Proxy.setWifiApConfiguration(IWifiManager.java:2288)
        at android.net.wifi.WifiManager.setWifiApConfiguration(WifiManager.java:2910)
        at java.lang.reflect.Method.invoke(Native Method) 
        at org.odk.share.controller.WifiHotspotHelper.setWifiConfig(WifiHotspotHelper.java:110) 
        at org.odk.share.controller.WifiHotspotHelper.disableHotspot(WifiHotspotHelper.java:133) 
        at org.odk.share.activities.SendActivity.onCreate(SendActivity.java:119) 
        at android.app.Activity.performCreate(Activity.java:7258) 
        at android.app.Activity.performCreate(Activity.java:7249) 
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1222) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2927) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3059) 
        at android.app.ActivityThread.-wrap11(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1724) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:7000) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:441) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1408) 
2019-03-20 18:58:11.524 16900-16900/org.odk.share W/WifiManager: org.odk.share attempted call to setWifiApEnabled: enabled = false
2019-03-20 18:58:11.537 16900-16900/org.odk.share I/WifiManager: startLocalOnlyHotspot() setwifiap packageName : org.odk.share
2019-03-20 18:58:11.542 16900-16900/org.odk.share D/AndroidRuntime: Shutting down VM
2019-03-20 18:58:11.547 16900-16900/org.odk.share E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.odk.share, PID: 16900
    java.lang.RuntimeException: Unable to resume activity {org.odk.share/org.odk.share.activities.SendActivity}: java.lang.SecurityException: Location mode is not enabled.
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3805)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3845)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3065)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1724)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:7000)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:441)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1408)
     Caused by: java.lang.SecurityException: Location mode is not enabled.
        at android.os.Parcel.readException(Parcel.java:2029)
        at android.os.Parcel.readException(Parcel.java:1975)
        at android.net.wifi.IWifiManager$Stub$Proxy.startLocalOnlyHotspot(IWifiManager.java:2176)
        at android.net.wifi.WifiManager.startLocalOnlyHotspot(WifiManager.java:2602)
        at org.odk.share.activities.SendActivity.turnOnHotspot(SendActivity.java:352)
        at org.odk.share.activities.SendActivity.initiateHotspot(SendActivity.java:161)
        at org.odk.share.activities.SendActivity.startHotspot(SendActivity.java:153)
        at org.odk.share.activities.SendActivity.onResume(SendActivity.java:257)
        at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1363)
        at android.app.Activity.performResume(Activity.java:7440)
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3780)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3845) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3065) 
        at android.app.ActivityThread.-wrap11(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1724) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:7000) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:441) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1408) 
```



![videotogif_2019 03 20_19 41 13](https://user-images.githubusercontent.com/29946262/54691162-c963cc00-4b48-11e9-8926-d6f735c4ac55.gif)

--- a/skunkworks_crow/src/main/java/org/odk/share/activities/SendActivity.java
+++ b/skunkworks_crow/src/main/java/org/odk/share/activities/SendActivity.java
@@ -6,11 +6,13 @@ import android.content.ComponentName;
 import android.content.Context;
 import android.content.DialogInterface;
 import android.content.Intent;
+import android.location.LocationManager;
 import android.net.wifi.WifiConfiguration;
 import android.net.wifi.WifiManager;
 import android.os.Build;
 import android.os.Bundle;
 import android.os.Handler;
+import android.provider.Settings;
 import android.support.annotation.RequiresApi;
 import android.support.v7.app.AlertDialog;
 import android.support.v7.widget.Toolbar;
@@ -156,8 +158,23 @@ public class SendActivity extends InjectableActivity {
         }
     }
 
+    private boolean isGPSEnabled() {
+        LocationManager locationManager = (LocationManager) getSystemService(LOCATION_SERVICE);
+        boolean gpsEnabled = false;
+        if (locationManager != null) {
+            gpsEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
+        }
+
+        return gpsEnabled;
+    }
+
     private void initiateHotspot() {
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            if (!isGPSEnabled()) {
+                isHotspotInitiated = false;
+                showLocationAlertDialog();
+                return;
+            }
             turnOnHotspot();
         } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
             // In devices having Android version = 7, created hotspot having some issues with connecting to other devices.
@@ -174,10 +191,32 @@ public class SendActivity extends InjectableActivity {
             Intent intent = new Intent(getApplicationContext(), HotspotService.class);
             intent.setAction(HotspotService.ACTION_STATUS);
             startService(intent);
-            
+
             currentConfig = wifiHotspot.getCurrConfig();
             startSending();
         }
+    }
+
+    private void showLocationAlertDialog() {
+        AlertDialog.Builder builder = new AlertDialog.Builder(this);
+        builder.setMessage(R.string.location_settings_dialog);
+        builder.setPositiveButton(getString(R.string.settings), new DialogInterface.OnClickListener() {
+            @Override
+            public void onClick(DialogInterface dialog, int which) {
+                Intent intent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
+                startActivity(intent);
+            }
+        });
+
+        builder.setNegativeButton(getString(R.string.cancel), new DialogInterface.OnClickListener() {
+            @Override
+            public void onClick(DialogInterface dialog, int which) {
+                dialog.dismiss();
+                finish();
+            }
+        });
+        builder.setCancelable(false);
+        builder.show();
     }
 
     private void showAlertDialog() {
@@ -267,7 +306,7 @@ public class SendActivity extends InjectableActivity {
             intent.setAction(HotspotService.ACTION_STATUS);
             startService(intent);
             Timber.d("Started hotspot N");
-            currentConfig = wifiHotspot.getCurrConfig();            
+            currentConfig = wifiHotspot.getCurrConfig();
             startSending();
         }
     }

--- a/skunkworks_crow/src/main/res/values/strings.xml
+++ b/skunkworks_crow/src/main/res/values/strings.xml
@@ -129,4 +129,5 @@
     <string name="default_odk_destination_dir">\/sdcard\/odk</string>
     <string name="title_odk_destination_dir">ODK Destination Directory</string>
     <string name="odk_destination_dir_error">The destination should not be empty</string>
+    <string name="location_settings_dialog">Enable location from the settings.</string>
 </resources>

